machines: Add bootorder option

skobyda opened the pull request:
Boot order defines order, in which devices will be booted. You can select which specific devices should be bootable. Right now, only disks and networks are supported as bootable devices.

skobyda commented:
![screenshot from 2018-12-20 15-57-23](https://user-images.githubusercontent.com/42733240/50292156-b3bf0780-0470-11e9-8652-9aacf4734e7c.png)
Also as you can see, there might be device "other - unsupported device". Right now, cockpit-machines only support disks and networks, but also host devices and redirected devices can have boot order. So if cockpit encounters machine with these devices, it needs to recognize that there are other devices with boot order, but labels them as "other - unsupported device". You can change order of this unsupported device, but cannot unselect it from list of bootable devices. Hope you agree with this approach.

If you start changing boot order when machine is running, warning pops up:
![screenshot from 2018-12-20 15-57-53](https://user-images.githubusercontent.com/42733240/50292160-b883bb80-0470-11e9-87e5-7ed6e991c5a3.png)

There will be also warning after you save changes. Warning disappears once machine is shut down. 
![screenshot from 2018-12-20 15-58-11](https://user-images.githubusercontent.com/42733240/50292167-bb7eac00-0470-11e9-9a39-c35e6935ba67.png)

In case of some internal or libvirt error, this warning might pop up:
![screenshot from 2018-12-20 16-01-08](https://user-images.githubusercontent.com/42733240/50292177-bfaac980-0470-11e9-804a-d13b731343a0.png)

skobyda commented:
@garrett can you review design and CSS please? :)

KKoukiou commented:
@skobyda thanks for this nice feature! Before I go on inline code review some overview.
This other - unsupported device in my opinion should be either hidden from the list completely, or if it's possible show the device info and disable the checkbox. But showing unsupported device, without hint of what device it is, is not really helpful to the used.
IMO it's better to just hide these devices since the user will not be able to use them anyway.

garrett commented:
Hi @skobyda, @KKoukiou: I just returned from holiday! Happy new year to you both!

Editing the boot order is a great idea. Thanks for making this feature.

## Stepping back a little

So, let's "step back" and look at the task. What is someone trying to do? They want to set the primary boot and some fallbacks, which may be manually selected at boot time.

Most people would want to have have a primary, a secondary, and perhaps a tertiary choice. In addition, some people would (most likely) want an arbitrary number of partitions to be able to boot from.

### Option 1: Pivot; use dropdowns

We could have boot order slots and dropdowns for the partition sources. When one is selected, that option disappears or becomes unselectable in the others. Or (probably better) if something is that is previous selected is selected again, the the previous dropdown changes to unselected. Unselected dropdowns would get filtered out when saved.

The number of dropdowns should be up to the maximum number of partitions, defaulting to 3 (to show that multiples can be selected), and if the number of dropdowns has not reached the number of partitions, another dropdown (without a partition selected) below the others should be shown.

Dropdowns would be sorted and grouped by type (disk, network, other).

The automatic unselected dropdown being added is similar to how the dynamic workspaces in GNOME works.

### Option 2: Drag and drop, with fallbacks

Another implementation would be a list with use drag and drop for reordering. It would allow someone to quickly change the boot order.

PatternFly has some notes written on a drag and drop widget @ https://www.patternfly.org/pattern-library/forms-and-controls/drag-and-drop/#overview & some design details @ https://www.patternfly.org/pattern-library/forms-and-controls/drag-and-drop/#design — but I've looked around a lot and there doesn't seem to be an implemented widget or official example code.

Of course, drag and drop would have to be able to be used in an accessible manner too. (It would need a way to adjust it with the keyboard.)

There's even an HTML 5 API for this drag and drop! However, the API:
1. [Is well supported by desktop browsers... but doesn't support mobile](https://caniuse.com/#feat=dragndrop).
   - [Touch is also a problem with ChromeOS touch input](https://bugs.chromium.org/p/chromium/issues/detail?id=915343).
2. Has keyboard support in the spec, but not in browsers.
   - [Firefox bug 504533](https://bugzilla.mozilla.org/show_bug.cgi?id=504533)
   - [WebKit but 27339](https://bugs.webkit.org/show_bug.cgi?id=27339)

For the cases where drag and drop to not work, we'd need to have a fallback. Tabbing to or clicking a number which represents order would allow someone to type in a new number to move the row to that location (and reorder the rest). This would also work where drag-and-drop is implemented.

### Option 3: Buttons

Changing list order with up and down actions is really clunky, especially for a list this size. The click target (up and down arrows) would change with every action as the list would constantly reorder itself. 

If you have something like 10 items and the one you want at the topis at the bottom, you have to scroll down, click the up arrow, move your mouse cursor up a bit, click again, move it up a bit (and repeat several times) and also scroll up somewhere in the process a little too.

Add top and bottom to the up and down actions, and make them all a button group inside every row.

As currently implemented, the icons are just colored icons, it's not clear which can be clicked and where one could click them. Actual, visibly grouped buttons would be better.

Additionally, the change-the-number mentioned in the drag and drop section above could be implemented for specific number ordering.

---

## Summary & Conclusion

Ideally, we would have drag and drop with fallbacks.

However, drag and drop implementations have issues (see above, in option 2) we'd have to work around and drag and drop would be the most complex widgets mentioned in this comment. In addition, we don't have any other reorderable list I'm aware of in Cockpit. As a result, perhaps we want to consider option 1 (dropdowns) or extending the buttons in this PR with option 3 instead?

What do you think?

garrett commented:
BTW: I'm happy to clarify anything I wrote in the previous comment and make a mockup for it too, if you have any questions.

andreasn commented:
@garrett I would not like to rely on html behaviour where both webkit and firefox have serious issues, so that would rule out option 2. I also don't fully understand the dropdowns concept from option 1. I think it needs some designs to clarify the concepts.

In general I think this PR looks fine as a starting point and solves the problem. Maybe not in the most elegant way, but it works well enough for me to merge. We can file followup issues with a revised design, and of course make sure to do usability tests to shake out the worst design problems.

skobyda commented:
I kinda did overkill with the number of devices in my example pictures. Usually number of devices in VM is lower, so bootorder list is smaller and moving something from up to all the way down the list and vice versa should not be problem most of the time.

But i agree that we can do follow up issues where we discuss problems with long list.
Drag and drop was my initial thought too. Perhaps we can ask somebody at patternfly to implement drag and drops?
Also I think 1st option would be a little unintuitive, but i don't know if i understand properly what you mean in 1st option.

skobyda commented:
So I added recognition of host devices and redirected devices. 

Also I changed the arrow into button groups as you wanted.

![screenshot from 2019-02-06 19-25-42](https://user-images.githubusercontent.com/42733240/52364312-331a9180-2a45-11e9-8841-aa1f68d12253.png)

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2018 Red Hat, Inc.
```
2019

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import React from 'react';
```
I tend to categorize the import with new line, above the external ones and bellow the imports from our own repo. Then new line and import css. Can we keep this at least in machines?

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+
+const DeviceInfo = ({ descrClass, descr, valueClass, value }) => {
+    return (
+        <div>
```
If you just need an empry wrapper element and not specifically the div, you can use <React.Fragment>

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+        <div>
+            <span className={descrClass}>
+                <label className='control-label'>
+                    {descr}
```
It's a label - value pair here? If so let's unwrap it from the outside span, and do something like
````
<React.Fragment>
    <label className={labelClasses} htmlFor={itemId}> desc </label>
    <span id={itemId}> {value} </span>
</React.Fragment>
````

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+        }
+    };
+
+    if (device.type === "disk") {
```
I think switch case is better for this else if hell, but I am not strongly opinionated on this.

skobyda replied:
Yeah, with switch it looks cleaner. I also changed if-else statements for switch in helper and libvirt-common.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+            devices,
+        }))
+                .fail((exc) => {
+                    this.dialogErrorSet(_("Boot order settings could not be saved"), exc.message);
```
Also then there is a single liner function handler you don't need the curly brackets. Just remove the brackets and bring it to the previous line.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,301 @@
+            vm,
+            devices,
+        }))
+                .fail((exc) => {
```
When there is single argument I tend to not wrap it in parenthesis.

skobyda commented:
Fixed it accordingly to @KKoukiou comments.

skobyda commented:
Also there is error, which fails test I wrote for bootorder. I suspect it is caused by patternfly, but I also might be wrong:
```
Warning: React does not recognize the `initExpanded` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `initexpanded` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in div (created by ListViewGroupItem)
    in ListViewGroupItem (created by ListViewItem)
    in ListViewItem (created by DeviceRow)
    in DeviceRow (created by BootOrderModal)
    in div (created by ListView)
    in ListView (created by BootOrderModal)
    in div (created by BootOrderModal)
    in div (created by ModalBody)
    in ModalBody (created by BootOrderModal)
    in div (created by CustomModalDialog)
    in div (created by CustomModalDialog)
    in div (created by CustomModalDialog)
    in CustomModalDialog (created by Modal)
    in Transition (created by Fade)
    in Fade (created by DialogTransition)
    in DialogTransition (created by Modal)
    in RefHolder (created by Modal)
    in div (created by Modal)
    in Portal (created by Modal)
    in Modal (created by Modal)
    in Modal (created by BootOrderModal)
    in BootOrderModal (created by VmOverviewTabLibvirt)
    in div (created by VmOverviewTabLibvirt)
    in VmOverviewTabLibvirt (created by ListingRow)
    in div (created by ListingRow)
    in td (created by ListingRow)
    in tr (created by ListingRow)
    in tbody (created by ListingRow)
    in ListingRow (created by Vm)
    in Vm (created by HostVmsList)
    in table (created by Listing)
    in Listing (created by HostVmsList)
    in div (created by HostVmsList)
    in HostVmsList (created by App)
    in div (created by App)
    in App
```

KKoukiou commented:
We agreed on IRC that xml-js should not added as dependency thus adding need work.

garrett commented:
> ![screenshot from 2019-02-06 19-25-42](https://user-images.githubusercontent.com/42733240/52364312-331a9180-2a45-11e9-8841-aa1f68d12253.png)

Looks good! I think the vertical alignment is a bit odd though — perhaps we want it all aligned to the first baseline. (Mainly talking about the first row.)

I _thought_ lists align to the first baseline already, though... Weird. Is this a PatternFly list or Cockpit list? I thought both are supposed to align to the first baseline. (I copied the behavior from an example from PatternFly to retrofit our lists.)

It's probably a 1-line fix if we want to change it. Otherwise, it's not really a huge deal.

skobyda commented:
Since https://github.com/cockpit-project/cockpit/pull/11179 introduced nodeDevices without bringing any dependency, so this PR can unblocked now :).
I added some more parsing to nodeDevices XML parsing to get all properties, which bootorder needs to know about nodeDevices.

@garrett It's Patternfly. The vertical alignment was caused by small space which was given for "additional info" (vendor, product...). However I overrided Patternfly's css with these values:
```
.list-view-pf-description {
    flex: 1 0 30%;
    width: 30%;
}

.list-view-pf-additional-info {
    width: 70%;
}
```
So now instead of 50%-50% division, there is 30%-70%, so more space for additional info values. Do you agree with approach?
Here's how it looks now:
![Screenshot from 2019-03-18 13-54-24](https://user-images.githubusercontent.com/42733240/54531356-5cec9f80-4985-11e9-97ca-0efe7d5a4011.png)

garrett commented:
I see you can boot from a USB camera? :wink: I guess the list should be limited to devices that can be booted from.

It's looking nice; I'll have to check it out and inspect the code a bit. Otherwise, I'm happy to see this coming along.

I do think the up/down UI isn't optimal, but it's fine enough for now (for this PR)... anything else would require a bit more work and might possibly be buggy (especially drag and drop)... and the list will probably not be _too_ long in most cases.

skobyda commented:
Ups, I don't know how I didn't notice incorrect term "hosted device". Fixed it to "host device".

skobyda commented:
@garrett Alright, example of webcam in my screenshot is stupid and doesn't have use case. However problem is, webcam is host device. And there are many other host devices which user might want to boot from. But I asked guys at libvirt, and they confirmed there is no way for us to figure out which host device does it make sense to boot from and which doesn't (like webcam) :(.
So there is no way to limit or filter host devices. I can only forbid every host device, but as I mentioned, booting from host devices does have use cases, so I would prefer not to forbid them.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+
+    const addOptional = (additionalInfo, value, descr) => {
+        if (value) {
+            additionalInfo.push(<DeviceInfo
```
Please put the <DeviceInfo in next row and there is not need to split into seperate lines the args any more.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+                addOptional(additionalInfo, nodeDev.capability.type.id, _("Type ID"));
+                break;
+            }
+            }
```
Do you want to add some default here with some debugLog that it's not handled? Or you handle everything?

skobyda replied:
According to https://libvirt.org/formatdomain.html#elementsHostDev and https://github.com/libvirt/libvirt/blob/master/docs/schemas/domaincommon.rng
these 5 types should be all there is.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+                break;
+            }
+            }
+            break;
```
This break is outside of the switch case.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+    }
+    }
+
+    let upArrow;
```
These are constants.

skobyda replied:
I don't know if I get what do you mean. I think declaring const variables without assignment would return error.

KKoukiou replied:
How you do it now you can't use constant because you have the assignment inside the if statement. You can use if you make in one liner with conditional assignment. Feel free to leave it as is. Didn't notice before that the assignment is in conditional.

skobyda replied:
That would be quite long one liner. I feel like this it looks nicer.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+        upArrow = (<Button onClick={moveUp}><Icon id={`${idPrefix}-up`} type="fa" name="angle-up" /></Button>);
+
+    if (downDisabled)
+        downArrow = (<Button disabled><Icon id={`${idPrefix}-down`} type="fa" name="angle-down" /></Button>);
```
Do you need parenthesis for all these assignments of buttons?

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+        super(props);
+        this.state = {
+            devices: getBootOrderDevices(props.vm),
+            dialogError: undefined,
```
No need to initialize variables to undefined. It's the default value.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+    move(d, direction) {
+        let devices = this.state.devices;
+
+        let index = devices.indexOf(d);
```
constant

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+        this.setState({ devices: devices });
+    }
+
+    move(d, direction) {
```
Can we rename d here and in all function that it's used with the same name to something more self explanatory?

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+
+        let index = devices.indexOf(d);
+        let tmp = devices[index + direction];
+        devices[index + direction] = devices[index];
```
You are writing directly on the state here. This is discouraged in react. https://reactjs.org/docs/state-and-lifecycle.html#do-not-modify-state-directly

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+        const { nodeDevices, vm } = this.props;
+        const idPrefix = vmId(vm.name) + '-order-modal';
+
+        function footerWarningHelper(device, index) {
```
Can you please start using some explanation regarding what functions do? In this one it's not so hard to understand indeed, but it's would be easier for review and future developement changes if we had some docstrings.
I started using this format recently. What do you think?

https://github.com/cockpit-project/cockpit/blob/master/pkg/machines/components/create-vm-dialog/pxe-helpers.js#L72

skobyda replied:
Sure, Doxygen style is good, and I already have some experience with that.

KKoukiou commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+            <Modal id={`${idPrefix}-window`} show onHide={this.close} className='boot-order'>
+                <Modal.Header>
+                    <Modal.CloseButton onClick={this.close} />
+                    <Modal.Title> {`${vm.name} Boot Order`} </Modal.Title>
```
This needs translation.

KKoukiou commented on pkg/machines/components/vmOverviewTabLibvirt.jsx:
```diff
@@ -64,19 +77,37 @@ class VmOverviewTabLibvirt extends React.Component {
+                // number of bootable devices has changed
+                return true;
+            } else {
+                // only order of devices has changes
```
It might be also possible that nothing has changed no?

KKoukiou replied:
In matters of bootable devices.

skobyda replied:
Yeah right. But if nothing changed, then !every() will return false, which is correct.
Comment in my code is misleading, I will change that.

KKoukiou commented on pkg/machines/helpers.js:
```diff
@@ -375,3 +377,164 @@ export function timeoutedPromise(promise, delay, afterTimeoutHandler, afterTimeo
+    return devices;
+}
+
+export function getSortedBootOrderDevices(vm) {
```
Please explain the logic behind this sorting. Either inline with comments either above the function definition.

garrett commented:
I've adjusted the layout in https://github.com/cockpit-project/cockpit/pull/10863/commits/06c1a59165714d72a98c2ba25beeefd631b9e816.

Notable changes:

1. Works in mobile sizes
2. Aligned to baseline
3. Hides actions for unchecked items
4. Larger tolerance before list starts scrolling, allowing several more items

---

## Improvements I'd like to see

- [ ] Clicking on the row (with the exception of the actions) should select/deselect the item
  - There should be a larger target area, including the name and possibly the details
  - This can be done with an event on row that toggles the state unless actions or the checkbox are clicked... the checkbox can handle itself; the actions shouldn't affect the checked status
- [ ] Improve accessibility
  - There's no label associated with the checkbox. If it's not possible to add a label, then it should use an `aria-labelledby` attribute with the id to the element that describes it or an `aria-label` attribute as a last-case resort if there's no descriptive element.
- [ ] Make up/down actions apply only to the selected rows
  - If the checked item has a bunch of items between it and another checked item above, then clicking up should jump it above the nextmost selected item above it. This should work similarly for down too.

## Things in the code that should probably be fixed

- [ ] You cannot target non-form elements with `<label>`.
  - I looked into this yesterday @ https://github.com/cockpit-project/cockpit/pull/11407#discussion_r266373786 and came to the conclusion that either the targeted elements need to be `<output>` or the label needs to be changed.
  - In the case of the other PR, we had to drop inner labels anyway as you cannot nest labels. That doesn't apply here, however.

## Stuff that should be fixed, but not in this PR

- We should not use floats or order to move items out of document order — but this is from PatternFly, so I guess it's more of a PatternFly bug we probably cannot fix here :cry:

garrett commented:
Screenshots!

### Desktop
![Screenshot_2019-03-19 Virtual Machines - Sunny](https://user-images.githubusercontent.com/10246/54604538-460e8180-4a47-11e9-8823-b25cd5172644.png)

### Mobile
![Screen Shot 2019-03-19 at 13 02 27](https://user-images.githubusercontent.com/10246/54604543-49a20880-4a47-11e9-8218-fcabdeaf6ace.png)

### Mobile small
![Screen Shot 2019-03-19 at 13 03 28](https://user-images.githubusercontent.com/10246/54604628-8110b500-4a47-11e9-9bfa-4b022aa2bbd1.png)

skobyda commented:
> * If the checked item has a bunch of items between it and another checked item above, then clicking up should jump it above the nextmost selected item above it. This should work similarly for down too.

Hmmm, if we are going to discuss this issue, then there is also a similar problem when the item, which is surrounded by selected items, is unchecked. This item won't change position and stays between selected items.

So alternatively, we could do jumping of items upon checking/unchecking.

Example of **checking item**:
The item is not selected. User checks the item, but there are unselected items in between this item and closest selected item. So the item would jump below nextmost selected (between selected and unselected items).

Example of **unchecking item**:
The item is selected. User unchecks the item, but the item is surrounded by selected items. So the item would jump below nextmost selected (between selected and unselected items).

skobyda commented:
@garrett Ups, I missed the last line sating it would work similarly for down. But the main problem I can imagine is this:
Let's imagine user has list o 10 selected items. He unselects every other Item (lets say 1st item, 3rd, 5th, 7th, 9th). Because if he moves the last selected item (10th item), it would move below nextmost selected item (8th item), but there would still be 4 unselected items above (1st, 3rd, 5th and 7th item).
So this approach you suggested wouldn't actually help much, right? I don't know, perphaps I'm overthinking stuff, or I didn't understand your suggestion correctly.

KKoukiou commented:
@skobyda also please squash your boot order commit with garrets css adjustments. And fix the commit message, it has typo.

skobyda commented on pkg/machines/helpers.js:
```diff
@@ -375,3 +377,164 @@ export function timeoutedPromise(promise, delay, afterTimeoutHandler, afterTimeo
+    }
+
+    const disks = Object.values(vm.disks);
+    // we ignore the disks which have been loaded from os->boot, and load the rest of disks
```
> he boot element can be repeated multiple times to setup a priority list of boot devices to try in turn
Nope, I also tried it and you can actually add multiple disks in boot->order.

Hmm, but another thing which I have noticed now, if you add N number of disks to boot->order, libvirt doesn't check if VM actually has N disks defined in devices. So I need to check that, otherwise it can crash.

KKoukiou replied:
> `boot element can be repeated multiple times to setup a priority list of boot devices to try in turn`
> 

You can repeat it multiple times but for different device types. Like
``` xml
  <boot dev='hd'/>
  <boot dev='cdrom'/>
```

> Nope, I also tried it and you can actually add multiple disks in boot->order.

You can add multiple disks but of different type aka file, cdrom.
But we don't take that in mind. See comment above. 

> 
> Hmm, but another thing which I have noticed now, if you add N number of disks to boot->order, libvirt doesn't check if VM actually has N disks defined in devices. So I need to check that, otherwise it can lead to error.

Good point. Libvirt should check that I guess though.

Antique replied:
I did not check the code so I don't know what you are planning to use, but please don't use the old boot order configuration.  Cockpit should be able to understand and parse the old <boot> element in the <os> element but it should never create such XML and it should always use the new way of per-device <boot> element.

The old way has a lot of issues especially if domain has a lot of disks and they are using different bus.

skobyda replied:
@Antique Exactly, right now boot order in this PR can read both old way and new way, but when you configure and save, it will always generate "new way XML".

Problem I mentioned was that if cockpit reads old boot order (os->boot) which contains let's say 4 disks, but VM has only 1 disk defined, there will be internal error in cockpit, since libvirt doesn't check that. (I think it will create array of disks with 1 defined and 3 undefined values or something like that). But I will fix that.

skobyda replied:
> You can repeat it multiple times but for different device types. Like
> 
> ```
>   <boot dev='hd'/>
>   <boot dev='cdrom'/>
> ```
Weird, I just configure XML of my VM with this and it works fine:
```
    <boot dev='hd'/>
    <boot dev='hd'/>
    <boot dev='network'/>
    <boot dev='network'/>
```

skobyda commented on pkg/machines/helpers.js:
```diff
@@ -375,3 +377,164 @@ export function timeoutedPromise(promise, delay, afterTimeoutHandler, afterTimeo
+
+            if (boot.type === "disk" | boot.type === "cdrom" | boot.type === "fd") {
+                devices.push({
+                    device: Object.values(vm.disks)[diskCount++],
```
As I mentioned below, I do not skip first disk only, but it can skip multiple disks. And in this line of code, diskCount++ doesn't do any skipping, it's there to really just keep count. The skipping of disks comes later when the cycle for cycling disks starts from diskCount value.

It seems that libvirt itself sorts the disks in devices according to target.

KKoukiou replied:
> As I mentioned below, I do not skip first disk only, but it can skip multiple disks. And in this line of code, diskCount++ doesn't do any skipping, it's there to really just keep count. The skipping of disks comes later when the cycle for cycling disks starts from diskCount value.
> 

I understand it. What I said it that skipping might not be correct. Let's sat that have have :+1:   <boot dev='cdrom'/> in the os element.

Which means that the first chrom disk should be bootable. And let's asume that there is not other <boot> element. So diskCount == 1.

Then you do to devices section in XML, and let's say it's like:
``` xml
  <disk type='volume' device='disk'>
    <driver name='qemu' type='raw'/>
    <source pool='blk-pool0' volume='blk-pool0-vol0'/>
    <target dev='vda' bus='virtio'/>
  </disk>
  <disk type='block' device='cdrom'>
    <driver name='qemu' type='raw'/>
    <target dev='hdd' bus='ide' tray='open'/>
    <readonly/>
  </disk>
```
What your code does is mark the first disk (which is the vda) as bootable but it should not do such.. Correct me if I am wrong.

> It seems that libvirt itself sorts the disks in devices according to target.

skobyda replied:
Oh right, I misunderstood the problem here. Alright, so I kinda changed the approach. Instead, all disks and interfaces are copied into arrays.
If XML uses the old way for boot order (os->boot), we find corresponding device in those arrays, and then remove it from the array. So there are only devices without boot order left.

skobyda commented on pkg/machines/components/vm/bootOrderModal.jsx:
```diff
@@ -0,0 +1,319 @@
+const DOWN = 1;
+
+function getBootOrderDevices(vm) {
+    let devices = getSortedBootOrderDevices(vm.inactiveXML);
```
Well, the thing here is, when the machine is running and you are adjusting boot order and save it, you are adjusting inactive (offline) XML. So it makes sense to list devices from inactive XML.

Now imagine that the machine is running, and somebody edits boot order from outside cockpit (thru virt-manager or virsh or whatever). Because the machine is running, it won't propagate to live XML yet, but it will be in inactive XML, so we want to show to user what's the state in inactive XML.
Otherwise we will rewrite the boot order changes which were done from outside cockpit or it may cause failure from libvirt (e. g. two devices were assigned same value). or (if I change how we update xml for boot order) user after restart gets boot order which he didn't want.

But it's true that this way, we exclude non-permanent disks. Do you have any idea how to solve both of these problems at once?

KKoukiou replied:
Since we are updating only the offline XML non permanent disks will not exist there so no need to handle them. Thanks for clarification.

However, the fact that we are updating the offline XML only it's our own implementation constraint. If we used https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainUpdateDeviceFlags API instead of redefining the whole domain XML that would be possible.
And to be honest it's nicer to allow people to just change boot order -> reboot and get their changes than having them to shut off the VM completely.

Interaction with other tools like virt-manager can cause inconsistent state to the VM indeed but I would not really put so much effort to avoid such scenarios where some API call fails because the user was messing up with it from two different tools.

So all in all, I am ok with this getting inside as is. But I would like to perform updates to the live XML as well, when there is API for it, like in this case.

Antique replied:
We cannot do updates to the live XML because you cannot change boot order for running guest.  If you want to add new disk or any other bootable device and you want to boot from it you need to shutdown the guest completely.

skobyda replied:
Okey, so if we cannot change boot order in live config, it resolves this conversation. We can update only offline config.

skobyda commented on pkg/machines/components/vmOverviewTabLibvirt.jsx:
```diff
@@ -24,20 +24,28 @@ import VmOverviewTab, { commonTitles } from './vmOverviewTab.jsx';
     let bootOrder = _("No boot device found");
-    if (vm.bootOrder && vm.bootOrder.devices && vm.bootOrder.devices.length > 0) {
-        bootOrder = vm.bootOrder.devices.map(bootDevice => bootDevice.type).join(); // Example: network,disk,disk
+    let devices = getSortedBootOrderDevices(vm.inactiveXML);
```
Oh right I agree, here it actually makes sense to list **current** state of boot order in active XML of running XML, nevertheless if some changes were done to inactive XML.

KKoukiou commented:
@skobyda the ovirt tests are failing because they are re-using the code from the machines plugin and need some adjustment,
https://209.132.184.41:8493/logs/pull-10863-20190326-172121-d9ad4c4c-verify-fedora-29/log.html#185

Let me know if you need help with this, is should be fairly easy. Sorry that I noticed late.

KKoukiou commented:
> @skobyda the ovirt tests are failing because they are re-using the code from the machines plugin and need some adjustment,
> https://209.132.184.41:8493/logs/pull-10863-20190326-172121-d9ad4c4c-verify-fedora-29/log.html#185
> 
> Let me know if you need help with this, is should be fairly easy. Sorry that I noticed late.

@skobyda Actually I fixed it for you reposted. Also I fixed a linters error for the tests.

martinpitt commented:
This still needs a "Closes #10863" in the main (second) commit so that we can land this. Thanks!

martinpitt commented:
Thanks @skobyda, great work!
